### PR TITLE
Incomplete stops first

### LINKED ISSUE #30: Incomplete Separated from Complete Stops
Requested during PVTA meeting on April 30th, 2018:

Prioritize stops that haven't been serviced in the list of stops.

### PR DESCRIPTION
**Pull request by mboneil10**
Closes #30

### CODE REVIEW

**Comment by dfaulken** [comment]
We can ask in this week's meeting, but I was under the impression that a more explicit separation was requested.

**Comment by mboneil10** [comment]
Yeah, that makes more sense than what I wrote. I can fix it by adding headers. I'll do that now, and delete the headers if requested (hooray for Git).

File: app/views/bus_stops/by_route.haml [hunk 1]
```diff
@@ -3,9 +3,14 @@
 - if @stops.empty?
   No stops on this route
 - else
-  %ul
-    - @stops.each do |stop|
+  - @inc_and_comp.each_pair do |complete, stops|
```
Comment (mboneil10) [inline app/views/bus_stops/by_route.haml]: `@inc_and_comp` is equivalent to `@stops.group_by(&:completed?)`, so `@stops.empty?` is fine to check that `@inc_and_comp` is empty.

File: app/controllers/bus_stops_controller.rb [hunk 2]
```diff
@@ -11,7 +11,8 @@ def autocomplete
   def by_route
     @route = Route.find_by number: params.require(:number)
     if @route.present?
-      @stops = @route.bus_stops.order(:name)
+      @stops = @route.bus_stops.order :name
+      @inc_and_comp = @stops.group_by(&:completed?)
```
Comment (mboneil10) [inline app/controllers/bus_stops_controller.rb]: I didn't know what to name this. I could combine line 14 and 15 into `@stops = @route.bus_stops.order(:name).group_by(&:completed?)` and fix the view accordingly.

**Comment by mboneil10** [comment]
screen shot 2018-05-29 at 2 04 50 pm

File: app/controllers/bus_stops_controller.rb (hunk 2 above)
Comment (mboneil10) [reply app/controllers/bus_stops_controller.rb]: Great, same. I'll do that.

File: app/views/bus_stops/by_route.haml [hunk 3]
```diff
@@ -3,9 +3,14 @@
 - if @stops.empty?
   No stops on this route
 - else
-  %ul
-    - @stops.each do |stop|
+  - @stops.each_pair do |complete, stops|
+    - if complete
```
Comment (mboneil10) [inline app/views/bus_stops/by_route.haml]: Should I change this to `completed`?

Comment (dfaulken) [reply app/views/bus_stops/by_route.haml]: I think it's fine. You could remove some duplication by doing

```haml
%ul
  - if complete
    Completed
  - else
    Not Completed
  Stops
```

If you so wished.

As a side note, it seems odd from a semantic perspective to have the title of the list be inside the list itself.

Comment (mboneil10) [reply app/views/bus_stops/by_route.haml]: OH

Comment (mboneil10) [reply app/views/bus_stops/by_route.haml]: Omg that's what I wished for.

**Comment by mboneil10** [comment]
screen shot 2018-05-30 at 2 08 43 pm

**Comment by mboneil10** [comment]
UGH HOLD ON

**Comment by mboneil10** [comment]
Also, the fact that I didn't do incomplete stops first. What was I thinking?

**Comment by mboneil10** [comment]
O whoops. I close while I fix.

**Comment by mboneil10** [comment]
Back in business.

File: app/controllers/bus_stops_controller.rb [hunk 4]
```diff
@@ -11,7 +11,7 @@ def autocomplete
   def by_route
     @route = Route.find_by number: params.require(:number)
     if @route.present?
-      @stops = @route.bus_stops.order(:name)
+      @stops = @route.bus_stops.order(:completed).group_by(&:completed?)
```
Comment (mboneil10) [inline app/controllers/bus_stops_controller.rb]: Just to refrain from writing `.group_by{|stop| stop.completed? == false}`. Honestly don't know which one is better.

**Comment by mboneil10** [comment]
During the meeting on May 31st, it was decided that this idea is fine. A few changes were requested:
- reformat list items to be `Sequence Name (Stop Number)`
  - side note: `sequence` needs to be added first
- order by sequence instead of lexicographically

**Comment by dfaulken** [comment]
> side note: sequence needs to be added first

From where in Hastus will we pull this value?

File: app/controllers/bus_stops_controller.rb (hunk 4 above)
Comment (dfaulken) [reply app/controllers/bus_stops_controller.rb]: Why do you want the boolean to be backwards?

**Comment by sherson** [comment]
PVTA is going to modify the export to give us the sequence from Hastus.

**Comment by dfaulken** [comment]
So PVTA is taking ownership of the OIG for this project? Should it no longer be in the contrib directory?

Still curious where they will pull the sequence from.

File: app/controllers/bus_stops_controller.rb (hunk 4 above)
Comment (mboneil10) [reply app/controllers/bus_stops_controller.rb]: So incomplete stops appear first

**Comment by mboneil10** [comment]
Going to make a separate issue for the additions to this page.

File: app/controllers/bus_stops_controller.rb (hunk 4 above)
Comment (mboneil10) [reply app/controllers/bus_stops_controller.rb]: Because incomplete stops have a higher priority for users than completed stops.

**Pull request merged**